snapshot/snapshot.go

```diff
@@ -0,0 +1,188 @@
+package snapshot
+
+import (
+	"database/sql"
+
+	"github.com/flanksource/incident-commander/db"
+	"github.com/flanksource/incident-commander/utils"
+)
+
+type resource struct {
+	componentIDs []string
+	configIDs    []string
+	incidentIDs  []string
+}
+
+func (src *resource) merge(dst resource) {
+	src.componentIDs = append(src.componentIDs, dst.componentIDs...)
+	src.configIDs = append(src.configIDs, dst.configIDs...)
+	src.incidentIDs = append(src.incidentIDs, dst.incidentIDs...)
+}
+
+func (r *resource) dedup() {
+	r.componentIDs = utils.Dedup(r.componentIDs)
+	r.configIDs = utils.Dedup(r.configIDs)
+	r.incidentIDs = utils.Dedup(r.incidentIDs)
+}
+
+func (r *resource) dump(directory string) error {
+	// Dedup since there maybe repetitions
+	r.dedup()
+
+	err := dumpComponents(directory, r.componentIDs)
+	if err != nil {
+		return err
+	}
+
+	err = dumpConfigs(directory, r.configIDs)
+	if err != nil {
+		return err
+	}
+
+	err = dumpIncidents(directory, r.incidentIDs)
+	if err != nil {
+		return err
+	}
+
+	return utils.Zip(directory, directory+".zip")
+}
+
+func topologySnapshot(componentID string, related bool, directory string) error {
+	var resources resource
+	componentIDs := []string{componentID}
+	resources.componentIDs = append(resources.componentIDs, componentIDs...)
+	if related {
+		relatedResources, err := fetchRelatedIDsForComponent(componentIDs)
+		if err != nil {
+			return err
+		}
+		resources.merge(relatedResources)
+
+		relatedConfigResources, err := fetchRelatedIDsForConfig(relatedResources.configIDs)
+		if err != nil {
+			return err
+		}
+		resources.merge(relatedConfigResources)
+	}
+
+	return resources.dump(directory)
+}
+
+func incidentSnapshot(incidentID, directory string) error {
+	var resources resource
+	resources.incidentIDs = []string{incidentID}
+	return resources.dump(directory)
+}
+
+func configSnapshot(configID string, related bool, directory string) error {
+	var resources resource
+	configIDs := []string{configID}
+	resources.configIDs = append(resources.configIDs, configIDs...)
+	if related {
+		relatedResources, err := fetchRelatedIDsForConfig(resources.configIDs)
+		if err != nil {
+			return err
+		}
+		resources.merge(relatedResources)
+	}
+
+	return resources.dump(directory)
+}
+
+func fetchRelatedIDsForComponent(componentIDs []string) (resource, error) {
```

Comment:
Can we 

a) Move most of the SQL into functions like https://github.com/flanksource/incident-commander/pull/208/files#diff-1a2d4a2b04d8d47065438c2b9deee4a37df131c4b8dcf5076db38f42abbe4da1R13 e.g. `lookup_related_components(id, depth=0)`
Depth being defaulted to 0, and when doing > 0 doing the recursive call to get child components

b) Abstract the sql function calls in the db package e.g. LookupRelatedComponents(), LookupIncidentsByComponent()